pkg/router/template/template_helper.go

```diff
@@ -386,6 +386,115 @@ func parseIPList(list string) string {
+	// Before enclosing the value in single quotes, # represented a comment and resulted in preventing the following
+	// characters from being parsed.
+	var escapeFlag bool
+	var sanitizedVal string
+	appendCaptureGroup := true
+	for _, char := range val {
+		if char == '#' && !escapeFlag {
+			// If # is encountered and not escaped, stop processing the string.
+			// TRICKY: To maintain compatibility, set appendCaptureGroup to false because the previous functionality
+			// commented out the capture group, effectively removing it.
+			appendCaptureGroup = false
+			break
+		}
+		escapeFlag = char == '\\'
+		// Otherwise, add the character to the sanitizedVal.
+		sanitizedVal += string(char)
+	}
+	val = sanitizedVal
+
+	// Remove all double quotes unless preceded by \.
+	// Before enclosing the value in single quotes, " represented weak quoting of which the contents were not
+	// interpreted literally.
+	escapeFlag = false
+	sanitizedVal = ""
+	for _, char := range val {
+		if char == '"' && !escapeFlag {
+			// If " is not escaped, but don't add " to sanitizedVal.
+			continue
+		}
+		escapeFlag = char == '\\'
+
+		// Otherwise, add the character to the sanitizedVal.
+		sanitizedVal += string(char)
+	}
+	val = sanitizedVal
+
+	// Remove all single quotes ' unless preceded by \, then replace with '\''.
+	// Because the value is automatically enclosed in single quotes, if a user deliberately includes a single quote, it
+	// will cancel out the existing single quotes, potentially causing haproxy to interpret the values or resulting
+	// in a syntax error.
+	escapeFlag = false
+	sanitizedVal = ""
+	for _, char := range val {
+
+		if char == '\'' && !escapeFlag {
+			// If ' is not escaped, but don't add ' to sanitizedVal.
+			continue
+		} else if char == '\'' && escapeFlag {
+			// If ' is escaped, then add '\'' to sanitizedVal.
+			// TRICKY: Because our later operation removes single \, we should double escape \\ it, so it will be
+			// converted to \ later.
+			sanitizedVal += `'\\''`
+			continue
+		}
+		escapeFlag = char == '\\'
+
+		// Otherwise, add the character to the sanitizedVal.
+		sanitizedVal += string(char)
+	}
+	val = sanitizedVal
+
+	// Convert percent sign % to become double percent signs %% if the sequence contains an odd number of
+	// percent signs.
+	// In the past, we didn't enclose the rewrite target value in single quotes to ensure haproxy interprets the values
+	// literally. As a result, a single % produced a syntax error, and users had to use %% to represent a literal %. To
+	// ensure compatibility while resolving this problem, we must continue to replace single % with double %%, but
+	// leave double %% as is, so that existing users who manually addressed the issue are not broken. Furthermore, even
+	// sequences, such as %%%% or %%%%%% were valid and should be left untouched, while odd sequences such as %%% or
+	// %%%%% were invalid, and should be doubled to ensure every % is escaped.
+	re := regexp.MustCompile(`%+`)
```

Comment:
If this function will be called frequently then we should be somewhat wary of the time and cost to compile the RE. See: https://github.com/openshift/router/pull/268

Reply:
Nice catch. Might as well optimize it. Done.

Reply:
Using `cachedRegexpCompile` makes sense where the pattern is built at run-time, but in this case, you can just use a package variable to avoid the map lookup and error checking:
```go
var (
	// oneOrMorePercentSigns matches one or more literal percent signs.
	oneOrMorePercentSigns = regexp.MustCompile(`%+`)
)

Reply:
Oh I see the intention of `cachedRegexpComplie` now. Done.